proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+be garbage collected.
+
+This also implies that the host VM must provide a garbage collector for
+exceptions.  For host VMs that have garbage collection (such as JavaScript),
+this is not a problem.
+
+However, not all host VMs may have a garbage collector. For this reason,
+WebAssembly exceptions are designed to allow the use of reference counters to
+perform the the garbage collection in the host VM.
+
+To do this, WebAssembly exceptions are immutable once created, to avoid cyclic
+data structures that can't be garbage collected. It also means that exceptions
+can't be stored into linear memory. The rationale for this is twofold:
+
+* For security. Loads
+  and stores do not guarantee that the data read was of the same type as
+  stored. This allows spoofing of exception references that may allow a
+  WebAssembly module to access data it should not know in the host VM.
+  
+* The host VM does not know the layout of data in linear memory, so it can't
+  find places where exception references are stored.
+
+Hence, while an exception reference is a new first class type, this proposal
+disallows their usage in linear memory.
+
+A WebAssembly exception is created by the `except` instruction. Once an
+exception is created, you can throw it with the `throw` instruction. Thrown
+exceptions are handled as follows:
+
+1. They can be caught by a catch block in an enclosing try block of a function
+   body. The caught exception is pushed onto the stack.
+
+1. Throws not caught within a function body continue up the call stack, popping
+   call frames, until an enclosing try block is found.
+   
+1. If the call stack is exhausted without any enclosing try blocks, the host VM
+   defines how to handle the uncaught exception.
+
+### Exceptions
+
+An `exception` is an internal construct in WebAssembly that is maintained by the
+host. WebAssembly exceptions (as opposed to host exceptions) are defined by a
+new `exception section` of a WebAssembly module. The exception section is a list
+of exception types, from which exceptions can be created.
+
+Each exception type has a `type signature`. The type signature defines the list
+of values associated with the exception.
+
+Within the module, exception types are identified by an index into the
+[exception index space](#exception-index-space). This index is referred to as
+the `exception tag`. The `tagged exception type` is the corresponding
+exception type refered to by the exception tag.
+
+Exception types can be imported and exported by adding the appropriate entries
+to the import and export sections of the module. All imported/exported exception
+types must be named to reconcile exception tags between modules.
+
+Exception tags are used by:
+
+1. The `except` instruction which creates a WebAssembly instance of the
+   corresponding tagged exception type, and pushes a reference to it onto the
+   stack.
+
+2. The `if_except` instruction that queries an exception to see if it is an
```

Comment:
I like the name `if_except`. What about `exception.match` if we decide to go with `exception.*` prefixes? Although we should probably save this discussion for a future bikeshedding period.

Reply:
Tracking in #35